pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -409,11 +424,133 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
+                tasks = []
+                for _, release_info in release_infos.items():
+                    pullspec = release_info["image"]
+                    digest = release_info["digest"]
+                    sig_file = json_digest_sig_dir / f"{digest.replace(':', '=')}" / "signature-1"
+                    tasks.append(self._sign_json_digest(signatory, release_info["metadata"]["version"], pullspec, digest, sig_file))
+                for message_digest in message_digests:
+                    input_path = base_to_mirror_dir / message_digest
+                    if not input_path.is_file():
+                        raise IOError(f"Message digest file {input_path} doesn't exist or is not a regular file")
+                    sig_file = message_digest_sig_dir / f"{message_digest}.gpg"
+                    tasks.append(self._sign_message_digest(signatory, release_name, input_path, sig_file))
+                await asyncio.gather(*tasks)
+
+        self._logger.info("All artifacts have been successfully signed.")
+        self._logger.info("Publishing signatures...")
+        tasks = [
+            self._publish_json_digest_signatures(json_digest_sig_dir),
+            self._publish_message_digest_signatures(message_digest_sig_dir),
+        ]
+        await asyncio.gather(*tasks)
+        self._logger.info("All signatures have been published.")
+
+    async def _sign_json_digest(self, signatory: AsyncSignatory, release_name: str, pullspec: str, digest: str, sig_path: Path):
+        """ Sign a JSON digest claim
+        :param signatory: Signatory
+        :param pullspec: Pullspec of the payload
+        :param digest: SHA256 digest of the payload
+        :param sig_path: Where to save the signature file
+        """
+        self._logger.info("Signing json digest for payload %s with digest %s...", pullspec, digest)
+        sig_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(sig_path, "wb") as sig_file:
+            if self.runtime.dry_run:
+                self._logger.warning("[DRY RUN] Would have signed the requested artifact.")
+                return
+            await signatory.sign_json_digest(
+                product="openshift",
+                release_name=release_name,
+                pullspec=pullspec,
+                digest=digest,
+                sig_file=sig_file)
+
+    async def _sign_message_digest(self, signatory: AsyncSignatory, release_name, input_path: Path, sig_path: Path):
+        """ Sign a message digest
+        :param signatory: Signatory
+        :param input_path: Path to the message digest file
+        :param sig_path: Where to save the signature file
+        """
+        self._logger.info("Signing message digest file %s...", input_path.absolute())
+        sig_path.parent.mkdir(parents=True, exist_ok=True)
+        with open(input_path, "rb") as in_file, open(sig_path, "wb") as sig_file:
+            if self.runtime.dry_run:
```

Comment:
Same as above, I would rather move the `dry_run` check right at the beginning of the function